src/Microsoft.Tye.Hosting/DockerRunner.cs

```diff
@@ -79,6 +79,11 @@ public async Task StartAsync(Application application)
                         continue;
                     }
 
+                    if (string.Equals(binding.Protocol, "udp", StringComparison.InvariantCultureIgnoreCase))
```

Comment:
I'm pretty sure this will always throw by default when specifying UDP, not just for proxies. This check is checking the bindings of a service.

I think the main things to verify are:
1. If a single container is running, works with UDP
2. If a single project is running, works with UDP
3. If multiple containers are running, works with UDP
4. If multiple containers and projects are running (docker proxy enabled), works with UDP
5. If an ingress is enabled, works with UDP.

Reply:
The piece of code i added the check and exception in appear to be responsible for adding proxies for access from inside docker to services running outside of docker.

This will only throw for any service without DockerRunInfo and without IngressRunInfo, so I think only services with project or executable type ( ? ) for which it will set up a proxy in docker, so it won't throw when you bind a udp port for a container here.

I'll try and verify the scenarios you mention above.

Reply:
Did you?

Reply:
Apologies for the delay...

**Scenario 1**

> If a single container is running, works with UDP

docker container runs and port is bound @ `0.0.0.0:6831->6831/udp`

```yaml
name: scenario-1
services:
- name: jaeger-service
  image: jaegertracing/all-in-one:1.21
  bindings:
  - name: compact
    port: 6831
    protocol: udp
```

```
$ docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED          STATUS          PORTS                                                                                   NAMES
1182ed35355f   jaegertracing/all-in-one:1.21   "/go/bin/all-in-one-…"   26 seconds ago   Up 23 seconds   5775/udp, 5778/tcp, 14250/tcp, 14268/tcp, 6832/udp, 16686/tcp, 0.0.0.0:6831->6831/udp   jaeger-service_68510150-b
```

**Scenario 2**

> If a single project is running, works with UDP

Project starts fine. But I get an error in my web app "Unrecognized scheme in server address 'udp://localhost:5050'. Only 'http://' is supported.", this can be fixed with PR #916.


```yaml
name: scenario-2
services:
- name: webapplication1
  project: WebApplication1/WebApplication1.csproj
  bindings:
  - name: udp
    port: 5050
    protocol: udp
```

**Scenario 3**

> If multiple containers are running, works with UDP

I'm assuming here that by "multiple containers", you mean multiple replicas of a container. This will not work because it would need to proxy the incoming udp packages to the correct service, and no udp proxy has been implemented.

```yaml
name: scenario-3
services:
- name: jaeger-service
  image: jaegertracing/all-in-one:1.21
  replicas: 2
  bindings:
  - name: compact
    port: 6831
    protocol: udp
```

```
$ tye run
Loading Application Details...
Launching Tye Host...

[21:38:54 INF] Executing application from C:\Users\Tim\source\repos\WebApplication1\tye.yaml
[21:38:54 INF] Dashboard running on http://127.0.0.1:8000
[21:38:54 INF] Mapping external port 6831 to internal port(s) 59186, 59187 for jaeger-service binding compact
Drats! 'run' failed:
        Proxy does not support the udp protocol yet.
```

If by "multiple containers" you mean different containers, this works just fine:

```yaml
name: scenario-3b
services:
- name: jaeger-service
  image: jaegertracing/all-in-one:1.21
  bindings:
  - name: compact
    port: 6831
    protocol: udp
- name: jaeger-service-2
  image: jaegertracing/all-in-one:1.21
  bindings:
  - name: compact
    port: 6832
    protocol: udp
```

```
$ docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED          STATUS          PORTS                                                                                   NAMES
f66fbb9e6cec   jaegertracing/all-in-one:1.21   "/go/bin/all-in-one-…"   16 seconds ago   Up 14 seconds   5775/udp, 5778/tcp, 14250/tcp, 14268/tcp, 6831/udp, 16686/tcp, 0.0.0.0:6832->6832/udp   jaeger-service-2_6c6da33c-9
4c26836fb8dc   jaegertracing/all-in-one:1.21   "/go/bin/all-in-one-…"   16 seconds ago   Up 14 seconds   5775/udp, 5778/tcp, 14250/tcp, 14268/tcp, 6832/udp, 16686/tcp, 0.0.0.0:6831->6831/udp   jaeger-service_5235c5b2-8
```

**Scenario 4**

> If multiple containers and projects are running (docker proxy enabled), works with UDP

This won't work since the proxy doesn't support udp yet.

```yaml
name: scenario-4
services:
- name: webapplication1
  project: WebApplication1/WebApplication1.csproj
  replicas: 2
  bindings:
  - name: udp
    port: 5050
    protocol: udp
```

```
$ tye run
Loading Application Details...
Launching Tye Host...

[21:49:02 INF] Executing application from C:\Users\Tim\source\repos\WebApplication1\tye.yaml
[21:49:03 INF] Dashboard running on http://127.0.0.1:8000
[21:49:03 INF] Mapping external port 5050 to internal port(s) 59817, 59819 for webapplication1 binding udp
Drats! 'run' failed:
        Proxy does not support the udp protocol yet.
```

**Scenario 5**

> If an ingress is enabled, works with UDP.

Ingress works fine in the scenario below. The udp connections are not touched by ingres.

```yaml
name: scenario-5
services:
- name: webapplication1
  project: WebApplication1/WebApplication1.csproj
  args: --urls http://*:5000 # workaround for #916
  bindings:
  - name: http
    port: 5000
  - name: udp
    port: 5050
    protocol: udp
- name: webapplication2
  project: WebApplication2/WebApplication2.csproj
  args: --urls http://*:5001 # workaround for #916
  bindings:
  - name: http
    port: 5001
  - name: udp
    port: 5051
    protocol: udp
ingress:
- name: http
  bindings:
  - port: 8080
    protocol: http
  rules:
  - service: webapplication1
    path: /app1
  - service: webapplication2
    path: /app2
```

But when I add a docker container it won't work, because it tries to add a proxy container.

A comment in tye explains the reason for the proxy:
```cs
// Inject a proxy per non-container service. This allows the container to use normal host names within the
// container network to talk to services on the host
```

```yaml
name: scenario-5
services:
- name: jaeger-service
  image: jaegertracing/all-in-one:1.21
  bindings:
  - name: compact
    port: 6831
    protocol: udp
- name: webapplication1
  project: WebApplication1/WebApplication1.csproj
  args: --urls http://*:5000 # workaround for #916
  bindings:
  - name: http
    port: 5000
  - name: udp
    port: 5050
    protocol: udp
- name: webapplication2
  project: WebApplication2/WebApplication2.csproj
  args: --urls http://*:5001 # workaround for #916
  bindings:
  - name: http
    port: 5001
  - name: udp
    port: 5051
    protocol: udp
ingress:
- name: http
  bindings:
  - port: 8080
    protocol: http
  rules:
  - service: webapplication1
    path: /app1
  - service: webapplication2
    path: /app2
```

```
$ tye run
Loading Application Details...
Launching Tye Host...

[22:08:01 INF] Executing application from C:\Users\Tim\source\repos\WebApplication1\tye.yaml
[22:08:01 INF] Dashboard running on http://127.0.0.1:8000
[22:08:01 INF] Processing ingress rule: Path:/app1, Host:null, Service:webapplication1
[22:08:01 INF] Service webapplication1 is using (5000, http://localhost:5000/)
[22:08:01 INF] Processing ingress rule: Path:/app2, Host:null, Service:webapplication2
[22:08:01 INF] Service webapplication2 is using (5001, http://localhost:5001/)
[22:08:02 INF] Docker image jaegertracing/all-in-one:1.21 already installed
Drats! 'run' failed:
        Proxy does not support the udp protocol yet.
```

If the projects do not use udp (but the docker container still does) it works just fine because now there is no need to proxy udp
```yaml
name: scenario-5
services:
- name: jaeger-service
  image: jaegertracing/all-in-one:1.21
  bindings:
  - name: compact
    port: 6831
    protocol: udp
- name: webapplication1
  project: WebApplication1/WebApplication1.csproj
  bindings:
  - name: http
    port: 5000
- name: webapplication2
  project: WebApplication2/WebApplication2.csproj
  bindings:
  - name: http
    port: 5001
ingress:
- name: http
  bindings:
  - port: 8080
    protocol: http
  rules:
  - service: webapplication1
    path: /app1
  - service: webapplication2
    path: /app2
```

```
$ docker ps
CONTAINER ID   IMAGE                                   COMMAND                  CREATED          STATUS          PORTS                                                                                   NAMES
bbfd5f6e30c1   mcr.microsoft.com/dotnet/core/sdk:3.1   "dotnet Microsoft.Ty…"   19 seconds ago   Up 16 seconds                                                                                           webapplication2-proxy_f831c996-1
284767eceada   mcr.microsoft.com/dotnet/core/sdk:3.1   "dotnet Microsoft.Ty…"   19 seconds ago   Up 16 seconds                                                                                           webapplication1-proxy_f1f02015-8
43ee389c7bad   jaegertracing/all-in-one:1.21           "/go/bin/all-in-one-…"   19 seconds ago   Up 17 seconds   5775/udp, 5778/tcp, 14250/tcp, 14268/tcp, 6832/udp, 16686/tcp, 0.0.0.0:6831->6831/udp   jaeger-service_de537efe-4
```

---

It might be better to just skip udp bindings here instead of throwing an error, this allows single-replica project services with udp bindings to work alongside docker container services.

Reply:
Please let me know if I should remove this error and skip the binding instead.